### DISCUSSION TITLE
Design question: Order change API

### DISCUSSION THREAD

**Discussion by raphaelm** [pretix/pretix#2536] [Ideas]
Our big issue #654 "Writable API" has just one feature missing that we now want to add: API support for changing orders, like you can do in the backend.

To have a proper definition for this, changing an order in this context means:

* changing the product or variation of a position
* changing the subevent of a position
* changing the seat of a position
* changing the price of a position
* changing the tax rule of a position
* changing the membership used for a position
* cancelling a position (already implemented)
* adding a position (top-level or add-on)
* splitting the order
* changing the value of a fee
* cancelling a fee
* adding a fee
* regenerating the secret of a position
* re-calculating all taxes from product and tax rule data

In the backend, there are also the following behavior options:

* send an email to customer (or not)
* re-issue any invoices (or not)

Now I see two directions for the API design and are happy for opinions:

## Option A: Try to model as native REST methods

Many of these could be very nicely modeled as native REST methods. For example, the one option we already implemented is "cancelling a position" which you can currently do with a simple ``DELETE …/order/XYZ12/positions/23/``. Similarly, we could imlement changing the product/seat/price/taxrule/membership as ``PATCH …/order/XYZ12/positions/23/`` with the respective bodies and "adding a position" as ``POST …/order/XYZ12/positions/``. We'd need some new URLs for fees and some special endpoints for e.g. splitting the order, but for the majority of operations this should work.

**Upside:** This would be a very simple, clean, and easily understandable API design. It would just work as someone might expect.

**Downsides:**

- There are some combinations of operations that would be impossible this way. As a construted example, if you have two positions, you could never swap their seats since any order of the two operations would include a forbidden operation. You'd need a switch using three operations, that technically risks making the seat available for sale in between.

- Using the behavior options like "reissue invoice" or "send email" becomes much harder, since – to avoid mail and invocie spam – you'd need logic in the client to only set them on the *last* operation you are doing.

- You need to pay much attention to error handling as a client, since if you e.g. want to change all products in an order, the first operations could succeed and the last ones fail, which leaves you with a tough situation to communicate back to your user.

## Option B: Add a special RPC-style API

An alternative API could look like this:

```
PATCH …/order/XYZ12/positions/

{
  "operations": [
    {"position": 1, "action": "change_seat", "seat": "Main-45-5"},
    {"position": 2, "action": "change_seat", "seat": "Main-45-6"},
    {"position": 2, "action": "change_item", "item": 3, "variation": 5},
    {"position": 3, "action": "cancel"},
    {"action": "add_position", "item": 4, "variation": 5, "addon_to": 2},
    {"position": 5, "action": "split"},
  ],
  "send_email": true,
  "reissue_invoice": false
}
```

**Upside:** Atomic API that either fully fails or fully succeeds, easily allows "swap" operations, basically solves all of the downsides from above.

**Downside:** Harder to understand for new consumers, requires more documentation and explanation, not standard-REST-style.

Any opinions?

**Comment by wiffbi** [comment]
I generally like the REST-approach better. It offers a clear and understandable API for most of the usual operations. Regarding the downsides: 

- Would it be possible to offer a `PATCH …/order/XYZ12/positions/` endpoint to patch all or some positions of an order in one „transaction“?
- If reissuing invoices or mail tasks become a problem, could we buffer this on the server as in we delay the mail-task for e.g. 10s and only add another mail-task if none is existing for the order and operation? Might be overkill though.
- For changing more than one product we could offer the `PATCH …/order/XYZ12/positions/` endpoint as above?

**Reply by pc-coholic** [reply]
I am also very much in favor of the REST-API style with the addition of an optional transaction model. Just not sure if we would introduce a dedicated transaction start/end endpoint or use for example `X-Transaction-ID`/`X-Transaction-Commit`-headers.

**Reply by raphaelm** [reply]
> Would it be possible to offer a PATCH …/order/XYZ12/positions/ endpoint to patch all or some positions of an order in one „transaction“?

So this would basically mean "do both options" :(

> If reissuing invoices or mail tasks become a problem, could we buffer this on the server as in we delay the mail-task for e.g. 10s and only add another mail-task if none is existing for the order and operation? Might be overkill though.

No, I think that's a bad solution, since it then maybe randomly works as expected sometimes and sometimes not.

> I am also very much in favor of the REST-API style with the addition of an optional transaction model. Just not sure if we would introduce a dedicated transaction start/end endpoint or use for example X-Transaction-ID/X-Transaction-Commit-headers.

Oof. This would require so much state handling on *both* the server of the client. I fail to see how this is any better than my approach B, and I even think it would be **less** REST-style.

**Reply by raphaelm** [reply]
> Oof. This would require so much state handling on both the server of the client. I fail to see how this is any better than my approach B, and I even think it would be less REST-style.

To elaborate: This would require all calls during the transaction to be basically no-ops and just queued on the server. So when you'd first do a PATCH, then a GET, and then the COMMIT, the GET would still show the old data, which would be confusing. Any attempt to solve this would end up at "we need to build this on top of actual database transactions" which means "all your rest calls in the transaction would need to be routed to the same webserver", which kills our entire scalability and deployment model, or "we're building a full database engine just for this". I do not think we want any cross-request state.

**Reply by wiffbi** [reply]
I am not to happy with this `X-Transaction-ID` header either – I think every request should be stateless. Regarding RPC-style API: I see that it’s probably easier to implement especially on the server and also would not require as many requests. Regarding my PATCH-REST-appraoch, I thought about something like this – just give the things that changed:

```
PATCH …/order/XYZ12/positions/

{
    1: {
        "seat": "Main-45-5"
    },
    2: {
        "item": 3,
        "variation": 5,
        "seat": "Main-45-6"
    }
}
```

But I guess this not as easy to implement on the server. The other operations such as cancel, add and split would need their own REST-based requests, which creates overhead. Also there is not way to combine those operations to make them atomic … Hm, the more I think about it, the RPC-style API might be the better solution.

**Reply by raphaelm** [reply]
> Regarding RPC-style API: I see that it’s probably easier to implement especially on the server

Easy implementation on our server is a factor, but not the most important one. But isn't it also easier to implement on the client?

**Reply by wiffbi** [reply]
> Easy implementation on our server is a factor, but not the most important one. But isn't it also easier to implement on the client?

For any non-primitive use case: IMHO easier to use the RPC-style on the client as well!

**Reply by raphaelm** [reply]
@wiffbi @pc-coholic 
After some thinking I've come to the conclusion that "do it both" is unfortunately the only correct answer here :(

I've implemented the REST-style operations on order positions in #2552. Not sure if I will do the RPC-style operations in the same PR or separately. Also, I'm not sure if it is maybe good-enough to only to RPC-style for the OrderFee objects, since they don't have API URLs yet

**Reply by crackwitz** [reply]
I know this is from 2022, when none of this existed yet.
Currently, on 2023.8, I'm trying to use the API to both create a position as well as put addons under it. Is that even supported? I can't `addon_to:???` onto a position that has no ID yet.
The available docs don't explain how to handle this.
I should probably start an issue about it, right? I'm looking for existing issues, haven't seen one so far.

**Reply by raphaelm** [reply]
> I know this is from 2022, when none of this existed yet.
> Currently, on 2023.8, I'm trying to use the API to both create a position as well as put addons under it. Is that even supported? I can't `addon_to:???` onto a position that has no ID yet.
> The available docs don't explain how to handle this.
> I should probably start an issue about it, right? I'm looking for existing issues, haven't seen one so far.

I think you'd need to do that in multiple steps

**Reply by crackwitz** [reply]
Thanks for your response! The docs weren't entirely clear on the capabilities, but I *suspected* as much. The *control page* for editing order positions *also* needs me to do this in two steps. The order *creation* API lets me assign positionids so I can do it in one step.
